File: apps/mobile/lib/constants/slash_commands.dart
/// Slash command definitions and default commands for different agent types.
/// Based on the vicoa-web implementation for consistency.

/// Represents a slash command with its metadata
class SlashCommand {
  final String command;
  final String description;
  final String source; // 'default' or 'custom'
  /// 'command' or 'skill'. Lets a future Skills tab list skills separately;
  /// built-in defaults are plain commands unless flagged otherwise.
  final String kind;
  /// Text to insert on selection when it differs from `command` — Codex
  /// invokes skills with `$name` rather than `/name`. Null means insert the
  /// command itself.
  final String? insert;

  const SlashCommand({
    required this.command,
    required this.description,
    this.source = 'default',
    this.kind = 'command',
    this.insert,
  });

  bool get isSkill => kind == 'skill';

  /// Creates a SlashCommand from a map (typically from API response)
  factory SlashCommand.fromMap(Map<String, dynamic> map) {
    String name = map['name']?.toString() ?? map['command']?.toString() ?? '';
    // Ensure command starts with '/'
    if (!name.startsWith('/')) {
      name = '/$name';
    }

    final insertRaw = map['insert']?.toString();
    return SlashCommand(
      command: name,
      description: map['description']?.toString() ?? '',
      source: map['source']?.toString() ?? 'custom',
      kind: map['kind']?.toString() == 'skill' ? 'skill' : 'command',
      insert: (insertRaw != null && insertRaw.isNotEmpty) ? insertRaw : null,
    );
  }

  /// Converts to a map for serialization
  Map<String, dynamic> toMap() {
    return {
      'name': command,
      'description': description,
      'source': source,
      'kind': kind,
      if (insert != null) 'insert': insert,
    };
  }

  @override
  bool operator ==(Object other) {
    if (identical(this, other)) return true;
    return other is SlashCommand && other.command == command;
  }

  @override
  int get hashCode => command.hashCode;
}

/// Agent types that support different command sets
enum AgentType {
  claude,
  codex,
  opencode,
}

/// Default Claude Code slash commands
/// Source: Claude Code v2.1.163
/// Synced with vicoa-web/lib/constants/slash-commands.ts
const List<SlashCommand> _claudeCommands = [
  // Session / context
  SlashCommand(command: '/clear', description: 'Start a new conversation. Pass a name to label the previous one in /resume. Aliases: /reset, /new'),
  SlashCommand(command: '/compact', description: 'Free up context by summarizing the conversation. Optionally pass focus instructions'),
  SlashCommand(command: '/context', description: 'Visualize current context usage as a colored grid'),
  SlashCommand(command: '/recap', description: 'Generate a one-line summary of the current session on demand'),
  SlashCommand(command: '/copy', description: 'Copy the last assistant response to clipboard. Pass N to copy the Nth-latest response'),
  SlashCommand(command: '/export', description: 'Export the current conversation as plain text'),

  // Task / workflow
  SlashCommand(command: '/btw', description: 'Ask a quick side question without adding to the conversation'),
  SlashCommand(command: '/plan', description: 'Enter plan mode. Pass a description to start immediately'),
  SlashCommand(command: '/goal', description: 'Set a goal: Claude keeps working until the condition is met. No arg shows current goal'),
  SlashCommand(command: '/fork', description: 'Spawn a forked subagent that inherits the full conversation and works in the background'),
  SlashCommand(command: '/background', description: 'Detach the current session to run as a background agent. Monitor with claude agents. Alias: /bg'),
  SlashCommand(command: '/stop', description: 'Stop the current background session (transcript and worktree are kept)'),
  SlashCommand(command: '/tasks', description: 'List and manage background tasks. Alias: /bashes'),
  SlashCommand(command: '/loop', description: 'Run a prompt repeatedly while the session stays open. Omit interval for model self-pacing. Alias: /proactive'),
  SlashCommand(command: '/schedule', description: 'Create, update, list, or run routines on Anthropic-managed cloud infrastructure. Alias: /routines'),
  SlashCommand(command: '/batch', description: 'Skill. Orchestrate large-scale changes across a codebase in parallel across up to 30 independent units'),
  SlashCommand(command: '/autofix-pr', description: "Spawn a Claude Code on the web session that watches the current branch's PR and pushes fixes when CI fails or reviewers comment. Requires gh CLI"),
  SlashCommand(command: '/ultraplan', description: 'Draft a plan in an ultraplan session, review in your browser, then execute remotely or send back to your terminal'),

  // Code review / analysis
  SlashCommand(command: '/review', description: 'Review a pull request locally. For deeper cloud-based review, use /code-review ultra'),
  SlashCommand(command: '/code-review', description: 'Skill. Review the current diff for bugs and cleanups. Pass --fix to apply, --comment to post as PR comments, or ultra for deep cloud review'),
  SlashCommand(command: '/simplify', description: 'Skill. Review changed code for cleanup opportunities and apply fixes. Does not look for bugs — use /code-review for that'),
  SlashCommand(command: '/security-review', description: 'Analyze pending changes on the current branch for security vulnerabilities'),
  SlashCommand(command: '/ultrareview', description: 'Run a deep, multi-agent code review in a cloud sandbox. Alias for /code-review ultra'),
  SlashCommand(command: '/deep-research', description: 'Workflow. Fan out web searches on a question, fetch and cross-check sources, and synthesize a cited report'),

  // Development
  SlashCommand(command: '/run', description: "Skill. Launch and drive your project's app to see a change working in the running app. Requires v2.1.145+"),
  SlashCommand(command: '/verify', description: 'Skill. Confirm a code change does what it should by building, running, and observing the app. Requires v2.1.145+'),
  SlashCommand(command: '/debug', description: 'Skill. Enable debug logging for the current session and troubleshoot issues'),
  SlashCommand(command: '/fewer-permission-prompts', description: 'Skill. Scan transcripts for common read-only tool calls and add an allowlist to reduce permission prompts'),

  // Setup / configuration
  SlashCommand(command: '/init', description: 'Initialize project with a CLAUDE.md guide. Set CLAUDE_CODE_NEW_INIT=1 for interactive setup'),
  SlashCommand(command: '/add-dir', description: 'Add a working directory for file access during the current session'),
  SlashCommand(command: '/terminal-setup', description: 'Install Shift+Enter key binding for newlines'),
  SlashCommand(command: '/reload-plugins', description: 'Reload all active plugins to apply pending changes without restarting'),
  SlashCommand(command: '/reload-skills', description: 'Re-scan skill and command directories so new or changed skills become available. Added in v2.1.152'),
  SlashCommand(command: '/run-skill-generator', description: "Skill. Teach /run and /verify how to build and launch your project's app. Requires v2.1.145+"),

  // Information
  SlashCommand(command: '/help', description: 'Show help and available commands'),
  SlashCommand(command: '/insights', description: 'Generate a report analyzing your Claude Code sessions, including project areas, interaction patterns, and friction points'),
  SlashCommand(command: '/team-onboarding', description: 'Generate a team onboarding guide from your Claude Code usage history (past 30 days)'),

  // Navigation
  SlashCommand(command: '/exit', description: 'Exit the CLI. In an attached background session, this detaches and the session keeps running. Alias: /quit'),
];

/// Default Codex slash commands
/// Source: codex-rs/tui/src/slash_command.rs
const List<SlashCommand> _codexCommands = [
  SlashCommand(command: '/new', description: 'Start a new chat during a conversation'),
  SlashCommand(command: '/clear', description: 'Clear the terminal and start a new chat'),
  SlashCommand(command: '/compact', description: 'Summarize conversation to prevent hitting the context limit'),
  SlashCommand(command: '/init', description: 'Create an AGENTS.md file with instructions for Codex'),
  SlashCommand(command: '/plan', description: 'Switch to Plan mode'),
  SlashCommand(command: '/goal', description: 'Set or view the goal for a long-running task'),
  SlashCommand(command: '/raw', description: 'Toggle raw scrollback mode for copy-friendly terminal selection'),
  // SlashCommand(command: '/skills', description: 'Use skills to improve how Codex performs specific tasks'),
  SlashCommand(command: '/approve', description: 'Approve one retry of a recent auto-review denial'),
  // /status and /ps return output that must be rendered back to the user
  // SlashCommand(command: '/status', description: 'Show current session configuration and token usage'),
  // SlashCommand(command: '/ps', description: 'List background terminals'),
  SlashCommand(command: '/stop', description: 'Stop all background terminals'),
  SlashCommand(command: '/clean', description: 'Stop all background terminals (alias for /stop)'),
];

/// Default OpenCode TUI slash commands
/// Based on vicoa-web/lib/constants/slash-commands.ts
const List<SlashCommand> _opencodeCommands = [
  SlashCommand(command: '/new', description: 'Start a new session'),
  SlashCommand(command: '/compact', description: 'Compact session'),
  SlashCommand(command: '/share', description: 'Share session'),
  SlashCommand(command: '/unshare', description: 'Unshare session'),
  SlashCommand(command: '/exit', description: 'Exit the app'),
];

/// Get default slash commands for a specific agent type
List<SlashCommand> getDefaultCommands(AgentType agentType) {
  switch (agentType) {
    case AgentType.claude:
      return _claudeCommands;
    case AgentType.codex:
      return _codexCommands;
    case AgentType.opencode:
      return _opencodeCommands;
  }
}

/// Get default commands for a string-based agent type identifier
List<SlashCommand> getDefaultCommandsByString(String agentType) {
  final type = agentType.toLowerCase();
  if (type.contains('claude')) {
    return getDefaultCommands(AgentType.claude);
  } else if (type.contains('codex')) {
    return getDefaultCommands(AgentType.codex);
  } else if (type.contains('opencode')) {
    return getDefaultCommands(AgentType.opencode);
  }
  // Default to Claude commands for unknown types
  return getDefaultCommands(AgentType.claude);
}

/// Merge default and custom commands, with custom commands overriding defaults
/// Custom commands with the same name as default commands will replace them
List<SlashCommand> mergeCommands(
  List<SlashCommand> defaultCommands,
  List<SlashCommand> customCommands,
) {
  // Preserve the preset order from defaultCommands.
  final result = List<SlashCommand>.from(defaultCommands);
  final Map<String, int> indexByCommand = {};

  for (var i = 0; i < result.length; i++) {
    indexByCommand[result[i].command] = i;
  }

  for (final cmd in customCommands) {
    // Preserve kind/insert so skills stay distinguishable downstream.
    final custom = SlashCommand(
      command: cmd.command,
      description: cmd.description,
      source: 'custom',
      kind: cmd.kind,
      insert: cmd.insert,
    );

    if (indexByCommand.containsKey(cmd.command)) {
      // Replace existing command while keeping its position.
      final idx = indexByCommand[cmd.command]!;
      result[idx] = custom;
    } else {
      // Append new custom commands after defaults in received order.
      indexByCommand[cmd.command] = result.length;
      result.add(custom);
    }
  }

  return result;
}

File: apps/web/lib/constants/slash-commands.ts
/**
 * Default slash commands for different AI agents
 * These are built-in commands that are always available
 */

export interface SlashCommand {
  command: string;
  description: string;
  source?: 'default' | 'custom';
  /**
   * Whether this entry is a plain slash command or an installed skill. Built-in
   * defaults omit it (treated as 'command'); custom entries carry it so a
   * future Skills tab can list skills separately.
   */
  kind?: 'command' | 'skill';
  /**
   * Text to insert into the composer on selection when it differs from
   * `command` — Codex invokes skills with `$name` rather than `/name`.
   */
  insert?: string;
}

export type AgentType = 'claude' | 'codex' | 'opencode';

/**
 * Claude Code built-in slash commands
 * Source: Claude Code v2.1.163
 */
export const CLAUDE_CODE_COMMANDS: SlashCommand[] = [
  // --- Session / context ---
  { command: '/clear', description: 'Start a new conversation. Pass a name to label the previous conversation in /resume. Aliases: /reset, /new', source: 'default' },
  { command: '/compact', description: 'Free up context by summarizing the conversation. Optionally pass focus instructions', source: 'default' },
  { command: '/context', description: 'Visualize current context usage as a colored grid', source: 'default' },
  { command: '/recap', description: 'Generate a one-line summary of the current session on demand', source: 'default' },
  { command: '/copy', description: 'Copy the last assistant response to clipboard. Pass N to copy the Nth-latest response', source: 'default' },
  { command: '/export', description: 'Export the current conversation as plain text', source: 'default' },

  // --- Task / workflow ---
  { command: '/btw', description: 'Ask a quick side question without adding to the conversation', source: 'default' },
  { command: '/plan', description: 'Enter plan mode. Pass a description to start immediately', source: 'default' },
  { command: '/goal', description: 'Set a goal: Claude keeps working until the condition is met. No arg shows current goal; clear/stop removes it', source: 'default' },
  { command: '/fork', description: 'Spawn a forked subagent that inherits the full conversation and works in the background', source: 'default' },
  { command: '/background', description: 'Detach the current session to run as a background agent. Monitor with claude agents. Alias: /bg', source: 'default' },
  { command: '/stop', description: 'Stop the current background session (transcript and worktree are kept)', source: 'default' },
  { command: '/tasks', description: 'List and manage background tasks. Alias: /bashes', source: 'default' },
  { command: '/loop', description: 'Run a prompt repeatedly while the session stays open. Omit interval for model self-pacing. Alias: /proactive', source: 'default' },
  { command: '/schedule', description: 'Create, update, list, or run routines on Anthropic-managed cloud infrastructure. Alias: /routines', source: 'default' },
  { command: '/batch', description: 'Skill. Orchestrate large-scale changes across a codebase in parallel across up to 30 independent units', source: 'default' },
  { command: '/autofix-pr', description: 'Spawn a Claude Code on the web session that watches the current branch\'s PR and pushes fixes when CI fails or reviewers comment. Requires gh CLI', source: 'default' },
  { command: '/ultraplan', description: 'Draft a plan in an ultraplan session, review it in your browser, then execute remotely or send back to your terminal', source: 'default' },

  // --- Code review / analysis ---
  { command: '/review', description: 'Review a pull request locally. For deeper cloud-based review, use /code-review ultra', source: 'default' },
  { command: '/code-review', description: 'Skill. Review the current diff for bugs and cleanups. Pass --fix to apply, --comment to post as PR comments, or ultra for deep cloud review', source: 'default' },
  { command: '/simplify', description: 'Skill. Review changed code for cleanup opportunities and apply fixes (reuse, simplification, efficiency). Does not look for bugs — use /code-review for that', source: 'default' },
  { command: '/security-review', description: 'Analyze pending changes on the current branch for security vulnerabilities (injection, auth issues, data exposure)', source: 'default' },
  { command: '/ultrareview', description: 'Run a deep, multi-agent code review in a cloud sandbox. Alias for /code-review ultra', source: 'default' },
  { command: '/deep-research', description: 'Workflow. Fan out web searches on a question, fetch and cross-check sources, and synthesize a cited report', source: 'default' },

  // --- Development ---
  { command: '/run', description: 'Skill. Launch and drive your project\'s app to see a change working in the running app. Requires v2.1.145+', source: 'default' },
  { command: '/verify', description: 'Skill. Confirm a code change does what it should by building, running, and observing the app. Requires v2.1.145+', source: 'default' },
  { command: '/debug', description: 'Skill. Enable debug logging for the current session and troubleshoot issues', source: 'default' },
  { command: '/fewer-permission-prompts', description: 'Skill. Scan transcripts for common read-only tool calls and add an allowlist to reduce permission prompts', source: 'default' },

  // --- Setup / configuration ---
  { command: '/init', description: 'Initialize project with a CLAUDE.md guide. Set CLAUDE_CODE_NEW_INIT=1 for interactive setup including skills, hooks, and personal memory', source: 'default' },
  { command: '/add-dir', description: 'Add a working directory for file access during the current session', source: 'default' },
  { command: '/terminal-setup', description: 'Install Shift+Enter key binding for newlines', source: 'default' },
  { command: '/reload-plugins', description: 'Reload all active plugins to apply pending changes without restarting', source: 'default' },
  { command: '/reload-skills', description: 'Re-scan skill and command directories so new or changed skills become available. Added in v2.1.152', source: 'default' },
  { command: '/run-skill-generator', description: 'Skill. Teach /run and /verify how to build and launch your project\'s app. Requires v2.1.145+', source: 'default' },

  // --- Information ---
  { command: '/help', description: 'Show help and available commands', source: 'default' },
  { command: '/insights', description: 'Generate a report analyzing your Claude Code sessions, including project areas, interaction patterns, and friction points', source: 'default' },
  { command: '/team-onboarding', description: 'Generate a team onboarding guide from your Claude Code usage history (past 30 days)', source: 'default' },

  // --- Navigation ---
  { command: '/exit', description: 'Exit the CLI. In an attached background session, this detaches and the session keeps running. Alias: /quit', source: 'default' },

  // --- Commands that open interactive TUI (output not displayable in web) ---
  // { command: '/model', description: 'Select or change the AI model', source: 'default' },
  // { command: '/mcp', description: 'Manage MCP server connections and OAuth authentication', source: 'default' },
  // { command: '/config', description: 'Open the Settings interface (Config tab)', source: 'default' },
  // { command: '/memory', description: 'Edit CLAUDE.md memory files', source: 'default' },
  // { command: '/permissions', description: 'View or update permissions', source: 'default' },
  // { command: '/hooks', description: 'Manage hook configurations for tool events', source: 'default' },
  // { command: '/agents', description: 'Manage custom AI subagents for specialized tasks', source: 'default' },
  // { command: '/plugin', description: 'Manage Claude Code plugins', source: 'default' },
  // { command: '/login', description: 'Switch Anthropic accounts', source: 'default' },
  // { command: '/logout', description: 'Sign out from your Anthropic account', source: 'default' },
  // { command: '/status', description: 'Open Settings interface showing version and account info', source: 'default' },
  // { command: '/rewind', description: 'Rewind the conversation and/or code', source: 'default' },
  // { command: '/resume', description: 'Resume a conversation', source: 'default' },
  // { command: '/fast', description: 'Toggle fast mode (uses Claude Opus with faster output)', source: 'default' },
  // { command: '/workflows', description: 'List and manage background workflows', source: 'default' },
  // { command: '/doctor', description: 'Checks the health of your Claude Code installation', source: 'default' },
  // { command: '/feedback', description: 'Submit feedback about Claude Code', source: 'default' },
  // { command: '/install-github-app', description: 'Set up Claude GitHub Actions for a repository', source: 'default' },
  // { command: '/privacy-settings', description: 'View and update your privacy settings', source: 'default' },
  // { command: '/sandbox', description: 'Enable sandboxed bash tool with filesystem isolation', source: 'default' },
  // { command: '/statusline', description: "Set up Claude Code's status line UI", source: 'default' },
  // { command: '/vim', description: 'Enter vim mode for alternating insert and command modes', source: 'default' },
  // { command: '/usage', description: 'Show plan usage limits and rate limit status', source: 'default' },
  // { command: '/ide', description: 'Manage IDE integrations and show status', source: 'default' },
  // { command: '/output-style', description: 'Set the output style directly or from a selection menu', source: 'default' },
  // { command: '/claude-api', description: 'Skill. Load Claude API reference material for your project language', source: 'default' },

  // --- Removed commands ---
  // { command: '/pr-comments', description: 'Removed in v2.1.91. Ask Claude directly to view pull request comments instead', source: 'default' },
];

/**
 * OpenAI Codex built-in slash commands
 * Source: codex-rs/tui/src/slash_command.rs
 */
export const CODEX_COMMANDS: SlashCommand[] = [
  { command: '/new', description: 'Start a new chat during a conversation', source: 'default' },
  { command: '/clear', description: 'Clear the terminal and start a new chat', source: 'default' },
  { command: '/compact', description: 'Summarize conversation to prevent hitting the context limit', source: 'default' },
  { command: '/init', description: 'Create an AGENTS.md file with instructions for Codex', source: 'default' },
  { command: '/plan', description: 'Switch to Plan mode', source: 'default' },
  { command: '/goal', description: 'Set or view the goal for a long-running task', source: 'default' },
  { command: '/raw', description: 'Toggle raw scrollback mode for copy-friendly terminal selection', source: 'default' },
  // { command: '/skills', description: 'Use skills to improve how Codex performs specific tasks', source: 'default' },
  { command: '/approve', description: 'Approve one retry of a recent auto-review denial', source: 'default' },
  // /status and /ps return output that must be rendered back to the user
  { command: '/status', description: 'Show current session configuration and token usage', source: 'default' },
  { command: '/ps', description: 'List background terminals', source: 'default' },
  { command: '/stop', description: 'Stop all background terminals', source: 'default' },
  { command: '/clean', description: 'Stop all background terminals (alias for /stop)', source: 'default' },
];

/**
 * OpenCode TUI built-in slash commands
 * Source: https://opencode.ai/docs/tui#commands
 */
export const OPENCODE_COMMANDS: SlashCommand[] = [
  // { command: '/sessions', description: 'Switch sessions', source: 'default' },
  { command: '/new', description: 'Start a new session', source: 'default' },
  // { command: '/models', description: 'Switch model', source: 'default' },
  // { command: '/agents', description: 'Switch agent', source: 'default' },
  // { command: '/mcps', description: 'Toggle MCPs', source: 'default' },
  // { command: '/connect', description: 'Connect provider', source: 'default' },
  // { command: '/status', description: 'View status', source: 'default' },
  // { command: '/themes', description: 'Switch theme', source: 'default' },
  // { command: '/help', description: 'Open help', source: 'default' },
  // { command: '/editor', description: 'Open editor', source: 'default' },
  { command: '/compact', description: 'Compact session', source: 'default' },
  { command: '/share', description: 'Share session', source: 'default' },
  // { command: '/rename', description: 'Rename session', source: 'default' },
  // { command: '/timeline', description: 'Jump to message', source: 'default' },
  // { command: '/fork', description: 'Fork from message', source: 'default' },
  { command: '/unshare', description: 'Unshare session', source: 'default' },
  // { command: '/undo', description: 'Undo previous message', source: 'default' },
  // { command: '/redo', description: 'Redo last undo', source: 'default' },
  // { command: '/timestamps', description: 'Toggle timestamps', source: 'default' },
  // { command: '/thinking', description: 'Toggle thinking', source: 'default' },
  // { command: '/copy', description: 'Copy session transcript', source: 'default' },
  // { command: '/export', description: 'Export session transcript', source: 'default' },
  { command: '/exit', description: 'Exit the app', source: 'default' },
];

/**
 * Get default commands for a specific agent type
 */
export function getDefaultCommands(agentType: AgentType): SlashCommand[] {
  switch (agentType) {
    case 'claude':
      return CLAUDE_CODE_COMMANDS;
    case 'codex':
      return CODEX_COMMANDS;
    case 'opencode':
      return OPENCODE_COMMANDS;
    default:
      return [];
  }
}

/**
 * Merge default and custom commands, maintaining default order and appending custom commands
 */
export function mergeCommands(
  defaultCommands: SlashCommand[],
  customCommands: SlashCommand[]
): SlashCommand[] {
  const defaultCommandNames = new Set(defaultCommands.map((cmd) => cmd.command));

  // Start with all default commands in their original order
  const result = [...defaultCommands];

  // Append only custom commands that don't override default ones. Preserve the
  // custom entry's kind/insert so skills stay distinguishable downstream.
  customCommands.forEach((cmd) => {
    if (!defaultCommandNames.has(cmd.command)) {
      result.push({ ...cmd, source: 'custom' });
    }
  });

  return result;
}
